File: sdks/python/wrapper/src/thunkmetrc/wrapper/ratelimiter.py
import asyncio
import time
from typing import Optional, Dict

class RateLimiterConfig:
    def __init__(self, 
                 enabled: bool = False,
                 max_get_per_second_per_facility: float = 50,
                 max_get_per_second_integrator: float = 150,
                 max_concurrent_get_per_facility: int = 10,
                 max_concurrent_get_integrator: int = 30):
        self.enabled = enabled
        self.max_get_per_second_per_facility = max_get_per_second_per_facility
        self.max_get_per_second_integrator = max_get_per_second_integrator
        self.max_concurrent_get_per_facility = max_concurrent_get_per_facility
        self.max_concurrent_get_integrator = max_concurrent_get_integrator

class TokenBucket:
    def __init__(self, rate: float, capacity: float):
        self.rate = rate
        self.capacity = capacity
        self.tokens = capacity
        self.last_refill = time.monotonic()
        self.lock = asyncio.Lock()

    async def wait(self):
        async with self.lock:
            now = time.monotonic()
            elapsed = now - self.last_refill
            self.tokens = min(self.capacity, self.tokens + elapsed * self.rate)
            self.last_refill = now

            if self.tokens >= 1:
                self.tokens -= 1
                return
            
            # Wait time
            missing = 1.0 - self.tokens
            wait_time = missing / self.rate
        
        await asyncio.sleep(wait_time)
        await self.wait() # Recurse to re-check

class MetrcRateLimiter:
    def __init__(self, config: Optional[RateLimiterConfig] = None):
        if config is None:
            config = RateLimiterConfig()
        self.config = config
        
        self.integrator_rate = TokenBucket(config.max_get_per_second_integrator, config.max_get_per_second_integrator)
        self.integrator_sem = asyncio.Semaphore(config.max_concurrent_get_integrator)
        
        self.facility_rates: Dict[str, TokenBucket] = {}
        self.facility_sems: Dict[str, asyncio.Semaphore] = {}
        self.lock = asyncio.Lock()

    async def execute(self, facility: Optional[str], is_get: bool, op):
        if not self.config.enabled or not is_get:
            return await op()

        async with self.integrator_sem:
            async with self.get_facility_sem(facility):
                await self.integrator_rate.wait()
                if facility:
                    await self.get_facility_rate(facility).wait()
                
                # Retry loop
                while True:
                    try:
                        return await op()
                    except Exception as e:
                        if "429" in str(e): # Checking generic error string, assume client raises exceptions
                             # TODO: Parse retry after
                             await asyncio.sleep(1.0)
                             continue
                        raise e

    def get_facility_rate(self, facility: str) -> TokenBucket:
        if facility not in self.facility_rates:
             self.facility_rates[facility] = TokenBucket(self.config.max_get_per_second_per_facility, self.config.max_get_per_second_per_facility)
        return self.facility_rates[facility]

    def get_facility_sem(self, facility: Optional[str]) -> asyncio.Semaphore:
        if not facility:
             # Return dummy semaphore or context manager
             return asyncio.Semaphore(9999) 
        if facility not in self.facility_sems:
             self.facility_sems[facility] = asyncio.Semaphore(self.config.max_concurrent_get_per_facility)
        return self.facility_sems[facility]

File: sdks/thunkmetrc-rs/wrapper/src/ratelimiter.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{Mutex, Semaphore};
use std::time::{Duration, Instant};

#[derive(Clone)]
pub struct RateLimiterConfig {
    pub enabled: bool,
    pub max_get_per_second_per_facility: f64,
    pub max_get_per_second_integrator: f64,
    pub max_concurrent_get_per_facility: usize,
    pub max_concurrent_get_integrator: usize,
    pub max_retries: usize,
}

impl Default for RateLimiterConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            max_get_per_second_per_facility: 50.0,
            max_get_per_second_integrator: 150.0,
            max_concurrent_get_per_facility: 10,
            max_concurrent_get_integrator: 30,
            max_retries: 5,
        }
    }
}

pub struct TokenBucket {
    rate: f64,
    capacity: f64,
    tokens: f64,
    last_refill: Instant,
}

impl TokenBucket {
    pub fn new(rate: f64, capacity: f64) -> Self {
        Self {
            rate,
            capacity,
            tokens: capacity,
            last_refill: Instant::now(),
        }
    }

    fn refill(&mut self) {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_refill).as_secs_f64();
        self.tokens = (self.tokens + elapsed * self.rate).min(self.capacity);
        self.last_refill = now;
    }

    pub fn try_consume(&mut self, amount: f64) -> Option<Duration> {
        self.refill();
        if self.tokens >= amount {
            self.tokens -= amount;
            None
        } else {
            let missing = amount - self.tokens;
            Some(Duration::from_secs_f64(missing / self.rate))
        }
    }
}

pub struct MetrcRateLimiter {
    config: RateLimiterConfig,
    integrator_rate: Mutex<TokenBucket>,
    integrator_sem: Arc<Semaphore>,
    facility_rates: Mutex<HashMap<String, Arc<Mutex<TokenBucket>>>>,
    facility_sems: Mutex<HashMap<String, Arc<Semaphore>>>,
}

impl MetrcRateLimiter {
    pub fn new(config: Option<RateLimiterConfig>) -> Self {
        let config = config.unwrap_or_default();
        Self {
            integrator_rate: Mutex::new(TokenBucket::new(config.max_get_per_second_integrator, config.max_get_per_second_integrator)),
            integrator_sem: Arc::new(Semaphore::new(config.max_concurrent_get_integrator)),
            facility_rates: Mutex::new(HashMap::new()),
            facility_sems: Mutex::new(HashMap::new()),
            config,
        }
    }

    pub async fn execute<F, Fut, T>(&self, facility: Option<&str>, is_get: bool, op: F) -> Result<T, Box<dyn std::error::Error + Send + Sync>>
    where
        F: Fn() -> Fut,
        Fut: std::future::Future<Output = Result<T, Box<dyn std::error::Error + Send + Sync>>>,
    {
        if !self.config.enabled || !is_get {
            return op().await;
        }

        // 1. Integrator Semaphore
        let _sem_permit = self.integrator_sem.acquire().await.map_err(|e| format!("Semaphore closed: {:?}", e)).unwrap();

        // 2. Facility Semaphore
        let _fac_permit = if let Some(f) = facility {
             let sem = {
                 let mut sems = self.facility_sems.lock().await;
                 sems.entry(f.to_string())
                     .or_insert_with(|| Arc::new(Semaphore::new(self.config.max_concurrent_get_per_facility)))
                     .clone()
             };
             Some(sem.acquire_owned().await.map_err(|e| format!("Semaphore closed: {:?}", e)).unwrap())
        } else {
             None
        };

        // 3. Global Rate
        loop {
            let mut bucket = self.integrator_rate.lock().await;
            if let Some(wait) = bucket.try_consume(1.0) {
                 drop(bucket);
                 tokio::time::sleep(wait).await;
            } else {
                 break;
            }
        }

        // 4. Facility Rate
        if let Some(f) = facility {
            let bucket_arc = {
                let mut rates = self.facility_rates.lock().await;
                rates.entry(f.to_string())
                    .or_insert_with(|| Arc::new(Mutex::new(TokenBucket::new(self.config.max_get_per_second_per_facility, self.config.max_get_per_second_per_facility))))
                    .clone()
            };
            loop {
                 let mut bucket = bucket_arc.lock().await;
                 if let Some(wait) = bucket.try_consume(1.0) {
                      drop(bucket);
                      tokio::time::sleep(wait).await;
                 } else {
                      break;
                 }
            }
        }

        // 5. Retry Loop with max retries
        let mut retry_count = 0;
        loop {
             let res = op().await;
             match res {
                 Ok(v) => return Ok(v),
                 Err(e) => {
                      if retry_count >= self.config.max_retries {
                          return Err(e);
                      }
                      retry_count += 1;

                      // Check for ApiError to respect Retry-After
                      // e is Box<dyn Error + Send + Sync>
                      if let Some(api_err) = e.downcast_ref::<thunkmetrc_client::ApiError>() {
                          if api_err.status == reqwest::StatusCode::TOO_MANY_REQUESTS {
                               if let Some(retry_after) = api_err.headers.get("Retry-After") {
                                   if let Ok(val_str) = retry_after.to_str() {
                                       if let Ok(secs) = val_str.parse::<u64>() {
                                           tokio::time::sleep(Duration::from_secs(secs)).await;
                                           continue;
                                       }
                                   }
                               }
                               // Default 429 wait
                               tokio::time::sleep(Duration::from_secs(1)).await;
                               continue;
                          }
                          // Handle 5xx with exponential backoff
                          if api_err.status.is_server_error() {
                              tokio::time::sleep(Duration::from_millis(500)).await;
                              continue; 
                          }
                      }
                      
                      let msg = format!("{:?}", e);
                      if msg.contains("429") {
                           tokio::time::sleep(Duration::from_secs(1)).await;
                           continue;
                      }
                      return Err(e);
                 }
             }
        }
    }
}
